File: app/src/main/java/com/coryswainston/smart/dictionary/activities/MainActivity.java
package com.coryswainston.smart.dictionary.activities;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Rect;
import android.hardware.Camera;
import android.os.Bundle;
import android.os.StrictMode;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.constraint.ConstraintSet;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.HapticFeedbackConstants;
import android.view.MotionEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import com.coryswainston.smart.dictionary.R;
import com.coryswainston.smart.dictionary.fragments.DefinitionsFragment;
import com.coryswainston.smart.dictionary.services.DetectorProcessor;
import com.coryswainston.smart.dictionary.util.CameraUtils;
import com.coryswainston.smart.dictionary.util.Settings;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.vision.CameraSource;
import com.google.android.gms.vision.text.Text;
import com.google.android.gms.vision.text.TextRecognizer;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import static com.coryswainston.smart.dictionary.util.CameraUtils.selectSizePair;
import static com.google.android.gms.vision.CameraSource.CAMERA_FACING_BACK;

public class MainActivity
        extends AppCompatActivity
        implements DefinitionsFragment.OnFragmentInteractionListener {

    private static final String TAG = "MainActivity";
    private boolean requested = false;

    private static final int DESIRED_WIDTH = 1200;
    private static final int DESIRED_HEIGHT = 1200;

    private CameraSource cameraSource;
    private SurfaceView surfaceView;
    private boolean surfaceAvailable;

    private Button captureButton;
    private Button defineButton;
    private View loadingGif;
    private ConstraintLayout wrapper;

    private DefinitionsFragment definitionsFragment;

    private String detectedText;
    private List<Text> blocks;
    private String selectedWord;
    private MotionEvent activeEvent;
    private float ratio;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        StrictMode.VmPolicy.Builder builder = new StrictMode.VmPolicy.Builder();
        StrictMode.setVmPolicy(builder.build());

        wrapper = findViewById(R.id.wrapper);
        captureButton = findViewById(R.id.capture_button);
        defineButton = findViewById(R.id.define_button);
        defineButton.setScaleX(0);
        defineButton.setScaleY(0);
        surfaceView = findViewById(R.id.surfaceView);
        loadingGif = findViewById(R.id.loading_gif);
        loadingGif.setVisibility(View.GONE);

        surfaceAvailable = false;
        if (checkForPermissions(Manifest.permission.CAMERA,
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.INTERNET)) {
            surfaceView.getHolder().addCallback(new SurfaceCallback());
        }
    }

    private void adjustPreviewSize() {
        // determine what the camera size will be
        try {
            Camera testInstance = Camera.open();
            CameraUtils.SizePair sizePair = CameraUtils.selectSizePair(testInstance, DESIRED_WIDTH, DESIRED_HEIGHT);

            int adjustedWidth = sizePair.previewSize().getWidth();
            int adjustedHeight = sizePair.previewSize().getHeight();

            testInstance.release();

            // manually change surfaceView to fit camera size
            surfaceView.setLayoutParams(new ConstraintLayout.LayoutParams(adjustedWidth, adjustedHeight));

            int height = surfaceView.getHeight();
            int width = surfaceView.getWidth();

            ratio = adjustedHeight / height;

            ConstraintSet constraintSet = new ConstraintSet();
            constraintSet.clone(wrapper);

            constraintSet.connect(R.id.surfaceView,ConstraintSet.LEFT,R.id.wrapper,ConstraintSet.LEFT,0);
//        constraintSet.connect(R.id.surfaceView,ConstraintSet.RIGHT,R.id.wrapper,ConstraintSet.RIGHT,0);
            constraintSet.connect(R.id.surfaceView,ConstraintSet.TOP,R.id.instruction_main,ConstraintSet.BOTTOM,24);
//        constraintSet.connect(R.id.surfaceView,ConstraintSet.BOTTOM,R.id.capture_button,ConstraintSet.TOP,24);
            constraintSet.applyTo(wrapper);
        } catch (Exception e) {
            Log.e(TAG, e.toString());
        }
    }

    private class SurfaceCallback implements SurfaceHolder.Callback {
        @Override
        public void surfaceCreated(final SurfaceHolder surfaceHolder) {
            surfaceAvailable = true;
            adjustPreviewSize();
            try {
                setUpCamera();
            }
            catch (SecurityException e) {
                Log.e("surfaceCallback", e.toString());
            }
            catch (IOException e) {
                Log.e("surfaceCallback", e.toString());
            }

            View.OnClickListener captureText = new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                    loadingGif.setVisibility(View.VISIBLE);

                    Intent intent = new Intent(MainActivity.this, DefineActivity.class);
                    intent.putExtra("detections", detectedText);
                    startActivity(intent);
                }
            };
            captureButton.setOnClickListener(captureText);

            surfaceView.setOnTouchListener(new View.OnTouchListener() {
                private boolean buzzed = false;

                @Override
                public boolean onTouch(View v, MotionEvent event) {

                    switch (event.getActionMasked()) {
                        case MotionEvent.ACTION_DOWN:
                        case MotionEvent.ACTION_MOVE:
                            if (eventTouchesDefineButton(event)) {
                                defineButton.setBackground(getResources().getDrawable(R.drawable.define_button_pressed_bg));
                                if (!buzzed) {
                                    defineButton.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY);
                                    buzzed = true;
                                }
                                return true;
                            } else {
                                defineButton.setBackground(getResources().getDrawable(R.drawable.define_button_bg));
                                buzzed = false;
                            }

                            activeEvent = event;

                            float screenX = event.getX();
                            float screenY = event.getY();

                            float realX = screenX * ratio;
                            float realY = screenY * ratio;

                            for(Text block : blocks) {
                                if (block.getBoundingBox().contains((int) realX, (int) realY)) {
                                    String tempWord = selectedWord;
                                    selectedWord = trimText(block);
                                    if (tempWord == null || !tempWord.equals(selectedWord)) {
                                        bringUpDefineButton(event);
                                    }
                                    break;
                                }
                            }
                            return true;
                        case MotionEvent.ACTION_UP:
                            defineButton.animate().scaleX(0).scaleY(0).setDuration(300);
                            if (eventTouchesDefineButton(event)) {
                                defineButton.setBackground(getResources().getDrawable(R.drawable.define_button_bg));
                                buzzed = false;
                                defineButton.performClick();
                            } else {
                                activeEvent = null;
                                selectedWord = null;
                            }
                            return true;
                        default:
                            return false;
                    }
                }
            });

        }

        private void bringUpDefineButton(MotionEvent event) {
            defineButton.setText(selectedWord);
            float x = event.getX() + surfaceView.getLeft() - (float)defineButton.getWidth() / 2f;
            if (x < 0) {
                x = 0;
            }
            if (x > surfaceView.getRight() - defineButton.getWidth()) {
                x = surfaceView.getRight() - defineButton.getWidth();
            }
            float y = event.getY() + surfaceView.getTop() - (float)defineButton.getHeight() - 80;

            if (defineButton.getScaleX() == 0) {
                defineButton.setX(x);
                defineButton.setY(y);
                defineButton.setTransformationMethod(null);
                defineButton.animate().scaleX(1).scaleY(1).setDuration(300);
            } else {
                defineButton.animate().translationX(x);
                defineButton.animate().translationY(y);
            }
        }

        private boolean eventTouchesDefineButton(MotionEvent event) {
            if (defineButton.getVisibility() == View.VISIBLE && defineButton.getScaleX() == 1) {
                int[] coords = new int[2];
                defineButton.getLocationInWindow(coords);
                Rect buttonPos = new Rect(coords[0],
                        coords[1],
                        coords[0] + defineButton.getWidth(),
                        coords[1] + defineButton.getHeight());
                if (buttonPos.contains((int) event.getX() + surfaceView.getLeft(),
                        (int) event.getY() + surfaceView.getTop())) {
                    return true;
                }
            }
            return false;
        }
        
        @Override
        public void surfaceDestroyed(SurfaceHolder surface) {
            surfaceAvailable = false;
        }

        @Override
        public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        }
    }

    public void onDefineButtonClick(View v) {
        if (definitionsFragment == null) {
            definitionsFragment = DefinitionsFragment.newInstance(selectedWord);
        } else {
            definitionsFragment.addWord(selectedWord);
        }
        definitionsFragment.show(this);
        activeEvent = null;
        selectedWord = null;
    }

    private String trimText(Text block) {
        if (block == null || activeEvent == null) {
            return null;
        }
        String fullText = block.getValue();
        int width = block.getBoundingBox().width();
        int distanceToTouch = (int) (activeEvent.getX() * ratio) - block.getBoundingBox().left;
        float howFarInAreWe = distanceToTouch / width;
        int textLength = fullText.length();
        int approximateIndex = (int)(textLength * howFarInAreWe);
        if (!Character.isLetter(fullText.charAt(approximateIndex))) {
            for (int i = 1;
                 approximateIndex - i >= 0 || approximateIndex + i < fullText.length();
                 i++) {
                if (approximateIndex - i >= 0 && Character.isLetter(fullText.charAt(approximateIndex - i))) {
                    approximateIndex = approximateIndex - i;
                    break;
                }
                if (approximateIndex + i < fullText.length() && Character.isLetter(fullText.charAt(approximateIndex + i))) {
                    approximateIndex = approximateIndex + i;
                    break;
                }
            }
        }
        int startIndex = approximateIndex;
        int endIndex = approximateIndex;
        for (int i = approximateIndex; i < textLength && Character.isLetter(fullText.charAt(i)); i++) {
            endIndex++;
        }
        for (int i = approximateIndex; i > 0 && Character.isLetter(fullText.charAt(i - 1)); i--) {
            startIndex--;
        }

        return fullText.substring(startIndex, endIndex).toLowerCase();
    }

    private boolean checkForPermissions(String ... permissions) {
        List<String> permissionsToRequest = new ArrayList<>();

        for (String permission: permissions) {
            if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
                permissionsToRequest.add(permission);
            }
        }
        if (permissionsToRequest.isEmpty()) {
            return true;
        }

        ActivityCompat.requestPermissions(MainActivity.this,
                permissionsToRequest.toArray(new String[permissionsToRequest.size()]), 0);
        return false;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] results) {
        if (requestCode != 0) {
            return;
        }

        for (int i = 0; i < results.length; i++) {
            if (results[i] != PackageManager.PERMISSION_GRANTED && permissions[i] != null) {
                AlertDialog alertDialog = new AlertDialog.Builder(MainActivity.this).create();
                alertDialog.setTitle("Permissions needed");
                alertDialog.setMessage("This app requires camera permissions to function properly.");
                alertDialog.setButton(AlertDialog.BUTTON_POSITIVE, "OK",
                        new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.dismiss();
                                ActivityCompat.requestPermissions(MainActivity.this,
                                        new String[] {
                                            Manifest.permission.CAMERA,
                                            Manifest.permission.READ_EXTERNAL_STORAGE,
                                            Manifest.permission.INTERNET
                                        },
                                        0);
                            }
                        });
                alertDialog.show();
                return;
            }
        }
        surfaceView.getHolder().addCallback(new SurfaceCallback());
    }

    private void setUpCamera() throws IOException, SecurityException {

        TextRecognizer recognizer = new TextRecognizer.Builder(getApplicationContext()).build();
        recognizer.setProcessor(new DetectorProcessor().withCallback(new DetectorProcessor.Callback() {
            @Override
            public void callback(List<Text> blocks, String detectedText) {
                MainActivity.this.detectedText = detectedText;
                MainActivity.this.blocks = blocks;
                if (activeEvent != null) {
                    surfaceView.onTouchEvent(activeEvent);
                }
            }
        }));

        if (!recognizer.isOperational()) {
            Log.w(TAG, "not operational");
        }

        cameraSource = new CameraSource.Builder(this, recognizer)
                .setFacing(CAMERA_FACING_BACK)
                .setRequestedFps(0.1f)
                .setRequestedPreviewSize(DESIRED_WIDTH, DESIRED_HEIGHT)
                .setAutoFocusEnabled(true)
                .build();

        startCameraSource();
    }

    private void startCameraSource() throws IOException, SecurityException {
        int code = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(getApplicationContext());
        if (code != ConnectionResult.SUCCESS) {
            Toast.makeText(this, "Play Services Unavailable", Toast.LENGTH_SHORT).show();
        }

        if (surfaceAvailable) {
            Log.d(TAG, "surface was available");
            cameraSource.start(surfaceView.getHolder());
            Log.d(TAG, "started cameraSource");
        } else {
            Log.d(TAG, "surface wasn't available");
        }
    }

    /**
     * Restarts the camera.
     */
    @Override
    protected void onResume() {
        super.onResume();

        loadingGif.setVisibility(View.GONE);

        if (cameraSource != null) {
            try {
                startCameraSource();
            }
            catch (IOException e) {
                Log.e(TAG, e.toString());
            }
        }
    }

    /**
     * Stops the camera.
     */
    @Override
    protected void onPause() {
        super.onPause();
        if (cameraSource != null) {
            cameraSource.stop();
        }
    }

    /**
     * Releases the resources associated with the camera source, the associated detectors, and the
     * rest of the processing pipeline.
     */
    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (cameraSource != null) {
            cameraSource.release();
        }
    }

    /**
     * Called when the user hits the exit button in the corner.
     */
    @Override
    public void onDefinitionFragmentExit(View v) {
        if (definitionsFragment.onExit()) {
            definitionsFragment = null;
        }
    }

    /**
     * Called when the user hits the settings button.
     */
    public void onSettingsClick(View v) {
        Settings.showDialog(this);
    }

    @Override
    public void toggleWordEdit(View v) {
        definitionsFragment.toggleWordEdit(v);
    }

    @Override
    public void onTabClick(View v) {
        definitionsFragment.onTabClick(v);
    }

    @Override
    public void onGoogleSearch(View v) {
        definitionsFragment.onGoogleSearch();
    }

    @Override
    public void onWikipediaSearch(View v) {
        definitionsFragment.onWikipediaSearch();
    }

    @Override
    public void onWordsBackOrForward(View v) {
        definitionsFragment.onWordsBackOrForward(v);
    }

    @Override
    public void onBackPressed() {
        if (definitionsFragment != null) {
            definitionsFragment.remove();
            definitionsFragment = null;
        } else {
            super.onBackPressed();
        }
    }
}

File: app/src/main/java/com/coryswainston/smart/dictionary/activities/MainActivity.kt
package com.coryswainston.smart.dictionary.activities

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.StrictMode
import android.os.StrictMode.VmPolicy
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.HapticFeedbackConstants
import android.view.MotionEvent
import android.view.SurfaceHolder
import android.view.View
import android.view.View.OnTouchListener
import android.widget.Toast
import com.coryswainston.smart.dictionary.R
import com.coryswainston.smart.dictionary.fragments.DefinitionsFragment
import com.coryswainston.smart.dictionary.fragments.DefinitionsFragment.OnFragmentInteractionListener
import com.coryswainston.smart.dictionary.services.DetectorProcessor
import com.coryswainston.smart.dictionary.util.Settings
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.vision.CameraSource
import com.google.android.gms.vision.text.Text
import com.google.android.gms.vision.text.TextRecognizer
import kotlinx.android.synthetic.main.activity_main.*
import java.io.IOException
import java.util.*

class MainActivity : AppCompatActivity(), OnFragmentInteractionListener {
    private var cameraSource: CameraSource? = null
    private var surfaceAvailable = false

    private var definitionsFragment: DefinitionsFragment? = null

    private var detectedText: String? = null
    private var blocks: List<Text>? = null
    private var selectedWord: String? = null
    private var activeEvent: MotionEvent? = null
    private val ratio = 0.9f

    private lateinit var bgButtonNormal: Drawable
    private lateinit var bgButtonTouch: Drawable

    @Suppress("deprecation") // for resources.getDrawable
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        bgButtonNormal = resources.getDrawable(R.drawable.define_button_bg)
        bgButtonTouch = resources.getDrawable(R.drawable.define_button_pressed_bg)

        val builder = VmPolicy.Builder()
        StrictMode.setVmPolicy(builder.build())

        define_button.scaleX = 0f
        define_button.scaleY = 0f
        loading_gif.visibility = View.GONE

        if (checkForPermissions(Manifest.permission.CAMERA,
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.INTERNET)) {
            surfaceView.holder.addCallback(SurfaceCallback())
        }
    }

    private inner class SurfaceCallback : SurfaceHolder.Callback {
        override fun surfaceCreated(surfaceHolder: SurfaceHolder) {
            surfaceAvailable = true
            try {
                setUpCamera()
            } catch (e: SecurityException) {
                Log.e("surfaceCallback", e.toString())
            } catch (e: IOException) {
                Log.e("surfaceCallback", e.toString())
            }

            val captureText = View.OnClickListener {
                loading_gif.visibility = View.VISIBLE
                val intent = Intent(this@MainActivity, DefineActivity::class.java)
                intent.putExtra("detections", detectedText)
                startActivity(intent)
            }

            capture_button.setOnClickListener(captureText)
            surfaceView.setOnTouchListener(object : OnTouchListener {

                private var buzzed = false

                override fun onTouch(v: View, event: MotionEvent): Boolean {
                    return when (event.actionMasked) {
                        MotionEvent.ACTION_DOWN, MotionEvent.ACTION_MOVE -> {
                            if (eventTouchesDefineButton(event)) {
                                define_button.background = bgButtonTouch

                                if (!buzzed) {
                                    define_button.performHapticFeedback(HapticFeedbackConstants.VIRTUAL_KEY)
                                    buzzed = true
                                }
                                return true
                            } else {
                                define_button.background = bgButtonNormal
                                buzzed = false
                            }
                            activeEvent = event
                            val screenX = event.x
                            val screenY = event.y
                            val realX = screenX * ratio
                            val realY = screenY * ratio
                            for (block in blocks!!) {
                                if (block.boundingBox.contains(realX.toInt(), realY.toInt())) {
                                    val tempWord = selectedWord
                                    selectedWord = trimText(block)
                                    if (tempWord == null || tempWord != selectedWord) {
                                        bringUpDefineButton(event)
                                    }
                                    break
                                }
                            }
                            true
                        }
                        MotionEvent.ACTION_UP -> {
                            define_button.animate().scaleX(0f).scaleY(0f).duration = 300
                            if (eventTouchesDefineButton(event)) {
                                define_button.background = bgButtonNormal
                                buzzed = false
                                define_button.performClick()
                            } else {
                                activeEvent = null
                                selectedWord = null
                            }
                            true
                        }
                        else -> false
                    }
                }
            })
        }

        private fun bringUpDefineButton(event: MotionEvent) {
            define_button.text = selectedWord
            var x = event.x + surfaceView!!.left - define_button.width.toFloat() / 2f
            if (x < 0) {
                x = 0f
            }
            if (x > surfaceView!!.right - define_button.width) {
                x = surfaceView!!.right - define_button.width.toFloat()
            }
            val y = event.y + surfaceView!!.top - define_button.height.toFloat() - 80
            if (define_button.scaleX == 0f) {
                define_button.x = x
                define_button.y = y
                define_button.transformationMethod = null
                define_button.animate().scaleX(1f).scaleY(1f).duration = 300
            } else {
                define_button.animate().translationX(x)
                define_button.animate().translationY(y)
            }
        }

        private fun eventTouchesDefineButton(event: MotionEvent): Boolean {
            if (define_button.visibility == View.VISIBLE && define_button.scaleX == 1f) {
                val coords = IntArray(2)
                define_button.getLocationInWindow(coords)
                val buttonPos = Rect(coords[0],
                        coords[1],
                        coords[0] + define_button.width,
                        coords[1] + define_button.height)
                if (buttonPos.contains(event.x.toInt() + surfaceView!!.left,
                                event.y.toInt() + surfaceView!!.top)) {
                    return true
                }
            }
            return false
        }

        override fun surfaceDestroyed(surface: SurfaceHolder) {
            surfaceAvailable = false
        }

        override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {}
    }

    @Suppress("unused_parameter")
    fun onDefineButtonClick(v: View?) {
        if (definitionsFragment == null) {
            definitionsFragment = DefinitionsFragment.newInstance(selectedWord)
        } else {
            definitionsFragment!!.addWord(selectedWord)
        }
        definitionsFragment!!.show(this)
        activeEvent = null
        selectedWord = null
    }

    private fun trimText(block: Text?): String? {
        if (block == null || activeEvent == null) {
            return null
        }
        val fullText = block.value
        val width = block.boundingBox.width()
        val distanceToTouch = (activeEvent!!.x * ratio).toInt() - block.boundingBox.left
        val howFarInAreWe = distanceToTouch / width.toFloat()
        val textLength = fullText.length
        var approximateIndex = (textLength * howFarInAreWe).toInt()
        if (!Character.isLetter(fullText[approximateIndex])) {
            var i = 1
            while (approximateIndex - i >= 0 || approximateIndex + i < fullText.length) {
                if (approximateIndex - i >= 0 && Character.isLetter(fullText[approximateIndex - i])) {
                    approximateIndex -= i
                    break
                }
                if (approximateIndex + i < fullText.length && Character.isLetter(fullText[approximateIndex + i])) {
                    approximateIndex += i
                    break
                }
                i++
            }
        }
        var startIndex = approximateIndex
        var endIndex = approximateIndex
        run {
            var i = approximateIndex
            while (i < textLength && Character.isLetter(fullText[i])) {
                endIndex++
                i++
            }
        }
        var i = approximateIndex
        while (i > 0 && Character.isLetter(fullText[i - 1])) {
            startIndex--
            i--
        }
        return fullText.substring(startIndex, endIndex).toLowerCase(Locale.getDefault())
    }

    private fun checkForPermissions(vararg permissions: String): Boolean {
        val permissionsToRequest: MutableList<String> = ArrayList()
        for (permission in permissions) {
            if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
                permissionsToRequest.add(permission)
            }
        }
        if (permissionsToRequest.isEmpty()) {
            return true
        }
        ActivityCompat.requestPermissions(this@MainActivity,
                permissionsToRequest.toTypedArray(), 0)
        return false
    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>,
                                            results: IntArray) {
        if (requestCode != 0) {
            return
        }
        for (i in results.indices) {
            if (results[i] != PackageManager.PERMISSION_GRANTED) {
                val alertDialog = AlertDialog.Builder(this@MainActivity).create()
                alertDialog.setTitle("Permissions needed")
                alertDialog.setMessage("This app requires camera permissions to function properly.")
                alertDialog.setButton(AlertDialog.BUTTON_POSITIVE, "OK"
                ) { dialog, _ ->
                    dialog.dismiss()
                    ActivityCompat.requestPermissions(this@MainActivity, arrayOf(
                            Manifest.permission.CAMERA,
                            Manifest.permission.READ_EXTERNAL_STORAGE,
                            Manifest.permission.INTERNET
                    ),
                            0)
                }
                alertDialog.show()
                return
            }
        }
        surfaceView!!.holder.addCallback(SurfaceCallback())
    }

    @Throws(IOException::class, SecurityException::class)
    private fun setUpCamera() {
        val recognizer = TextRecognizer.Builder(applicationContext).build()
        recognizer.setProcessor(DetectorProcessor().withCallback { blocks, detectedText ->
            this@MainActivity.detectedText = detectedText
            this@MainActivity.blocks = blocks
            if (activeEvent != null) {
                surfaceView!!.onTouchEvent(activeEvent)
            }
        })
        if (!recognizer.isOperational) {
            Log.w(TAG, "not operational")
        }
        cameraSource = CameraSource.Builder(this, recognizer)
                .setFacing(CameraSource.CAMERA_FACING_BACK)
                .setRequestedFps(0.1f)
                .setRequestedPreviewSize(DESIRED_WIDTH, DESIRED_HEIGHT)
                .setAutoFocusEnabled(true)
                .build()
        startCameraSource()
    }

    @Throws(IOException::class, SecurityException::class)
    private fun startCameraSource() {
        val code = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(applicationContext)
        if (code != ConnectionResult.SUCCESS) {
            Toast.makeText(this, "Play Services Unavailable", Toast.LENGTH_SHORT).show()
        }
        if (surfaceAvailable) {
            Log.d(TAG, "surface was available")
            cameraSource!!.start(surfaceView!!.holder)
            Log.d(TAG, "started cameraSource")
        } else {
            Log.d(TAG, "surface wasn't available")
        }
    }

    /**
     * Restarts the camera.
     */
    override fun onResume() {
        super.onResume()
        loading_gif.visibility = View.GONE
        if (cameraSource != null) {
            try {
                startCameraSource()
            } catch (e: IOException) {
                Log.e(TAG, e.toString())
            }
        }
    }

    /**
     * Stops the camera.
     */
    override fun onPause() {
        super.onPause()
        if (cameraSource != null) {
            cameraSource!!.stop()
        }
    }

    /**
     * Releases the resources associated with the camera source, the associated detectors, and the
     * rest of the processing pipeline.
     */
    override fun onDestroy() {
        super.onDestroy()
        if (cameraSource != null) {
            cameraSource!!.release()
        }
    }

    /**
     * Called when the user hits the exit button in the corner.
     */
    override fun onDefinitionFragmentExit(v: View) {
        if (definitionsFragment!!.onExit()) {
            definitionsFragment = null
        }
    }

    /**
     * Called when the user hits the settings button.
     */
    @Suppress("unused_parameter")
    fun onSettingsClick(v: View?) {
        Settings.showDialog(this)
    }

    override fun toggleWordEdit(v: View) {
        definitionsFragment!!.toggleWordEdit(v)
    }

    override fun onTabClick(v: View) {
        definitionsFragment!!.onTabClick(v)
    }

    override fun onGoogleSearch(v: View) {
        definitionsFragment!!.onGoogleSearch()
    }

    override fun onWikipediaSearch(v: View) {
        definitionsFragment!!.onWikipediaSearch()
    }

    override fun onWordsBackOrForward(v: View) {
        definitionsFragment!!.onWordsBackOrForward(v)
    }

    override fun onBackPressed() {
        if (definitionsFragment != null) {
            definitionsFragment!!.remove()
            definitionsFragment = null
        } else {
            super.onBackPressed()
        }
    }

    companion object {
        private const val TAG = "MainActivity"
        private const val DESIRED_WIDTH = 1200
        private const val DESIRED_HEIGHT = 1200
    }
}